### ISSUE TITLE
planner: inaccurate estimated rows when DNF and CNF are nested together

### ISSUE THREAD

**Issue by qw4990** [pingcap/tidb#23192]
## Development Task
```
create table t (
    id int, 
    c int,
    index idx(id, c)
);

insert into t values (1, 1), (1, 1), (1, 1), (1, 1), (1, 1);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t values (2, 2);
analyze table t;
select count(*) from t; // 40 rows of (1, 1) and 1 row of (2, 2)

explain select * from t where
    (   id > 2
        or
        (id = 2 and c >= 2))
    and
    (   id < 2
        or
        (id = 2 and c <= 2));
+--------------------------+---------+-----------+---------------------------+----------------------------------------------------------------------------------------------------------------------------+
| id                       | estRows | task      | access object             | operator info                                                                                                              |
+--------------------------+---------+-----------+---------------------------+----------------------------------------------------------------------------------------------------------------------------+
| IndexReader_7            | 32.80   | root      |                           | index:Selection_6                                                                                                          |
| └─Selection_6            | 32.80   | cop[tikv] |                           | or(gt(test.t.id, 2), and(eq(test.t.id, 2), ge(test.t.c, 2))), or(lt(test.t.id, 2), and(eq(test.t.id, 2), le(test.t.c, 2))) |
|   └─IndexRangeScan_5     | 41.00   | cop[tikv] | table:t, index:idx(id, c) | range:[2,2], keep order:false                                                                                              |
+--------------------------+---------+-----------+---------------------------+----------------------------------------------------------------------------------------------------------------------------+
```
In the explain result, the access range is the correct value `[2, 2]` while its `estRows` is `41.00` which is too larger for this range.
I think its root cause is the way we deal with nested DNF and CNF is too naive.

**Comment by qw4990** [comment]
Do you have any idea about this? @eurekaka @time-and-fate 
We can fix issues like this case by case, but would you have any idea that can improve the ability of our planner to process nested DNF/CNF thoroughly?

**Comment by time-and-fate** [comment]
After investigation, this issue exists on the release-4.0 branch. On the master branch, it's been indirectly fixed by #18741.

The direct cause is that, on the release-4.0 branch, it will go into the condition branch below, but on the master branch, it won't.
https://github.com/pingcap/tidb/blob/aa82a038f1aa1739885592b57251eb77eaff9991/planner/core/logical_plans.go#L707-L711

The reason behind this is that on the master branch, the `ds.stats.RowCount` is less than 1 and on the release-4.0 branch, it's 41 * 0.8 ≈ 32. The difference is caused by changes in `Selectivity()` in #18741.

**Comment by time-and-fate** [comment]
And we're cherry-picking #18741 to release-4.0 now.
